Adding Cognito to the services that can be spun up in the CDK stack

bryan-harter opened the pull request.

bryan-harter committed:
Adding in context for the unit tests

bryan-harter committed:
Adding libraries to pyproject.toml

bryan-harter committed:
pip installs come from pyproject.toml now

bryan-harter committed:
Updating context to be passed into the function, rather than getting it twice

bryan-harter committed:
Keeping this branch up to date with the other tickets

bryan-harter committed:
Fixing a few things before the pull request

bryan-harter committed:
Adding a cognito utilities module

bryan-harter committed:
A few more fixes

bryan-harter committed:
Making changes to put cognito in a seperate stack

bryan-harter committed:
Fixing the docs, tests, and import errors

bryan-harter committed:
Merge branch 'development' into harter_cognito

# Conflicts:
#	.devcontainer/Dockerfile
#	.devcontainer/devcontainer.json
#	README.md
#	app.py
#	pyproject.toml
#	scripts/sds_api.py
#	sds_data_manager/lambda_code/SDSCode/download_query_api.py
#	sds_data_manager/lambda_code/SDSCode/queries.py
#	sds_in_a_box/SDSCode/upload_api.py
#	sds_in_a_box/sds_in_a_box_stack.py
#	tests/unit/test_sds_in_a_box_stack.py

bryan-harter committed:
Fixing many bad things from the merge

bryan-harter committed:
fixing the upload_api formatting issues

bryan-harter committed:
More merge fixing

bryan-harter committed:
A few more issues from the merge

bryan-harter committed:
More bug fixes from the merge

bryan-harter committed:
Removing the API verification stuff to make a smaller commit

bryan-harter committed:
Fixing the cognito unit tests

bryan-harter committed:
Attempting to fix some of the formatting

bryan-harter committed:
More playing around with the formatting

bryan-harter commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bryan-harter committed:
Fixing more formatting issues

bryan-harter commented:
pre-commit.ci autofix

bryan-harter committed:
Merge branch 'development' into harter_cognito_services

bryan-harter commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

greglucas commented on app.py:
```diff
@@ -1,14 +1,57 @@
 
+# Grab context from cdk synth and cdk deploy commands
 SDS_ID = app.node.try_get_context("SDSID")
+initial_user = app.node.try_get_context("initial_user")
+userpool_name = app.node.try_get_context("userpool_name")
+app_client_name = app.node.try_get_context("app_client_name")
+cognito_only = app.node.try_get_context("cognito_only")
```
This seems like it is adding a lot of complication and confusion for deployment. We'll now have to remember all of these options if we want to deploy something.

bryan-harter replied:
If we do put the cognito in a separate stack, how would we tell the data-management stack what the userpool/app_clients are? Would they be hard-coded?

greglucas commented on sds_data_manager/lambda_code/SDSCode/cognito_signup_message.py:
```diff
@@ -0,0 +1,80 @@
+            )
+            event["response"]["emailMessage"] = message
+
+            print(event)
```
```suggestion
```

greglucas commented on sds_data_manager/lambda_code/SDSCode/cognito_signup_message.py:
```diff
@@ -0,0 +1,80 @@
+            # link to the cognito domain.
+            subject = f"Science Data System {os.environ['SDS_ID']} sign up"
+            event["response"]["emailSubject"] = subject
+            event["response"]["smsMessage"] = "We don't use this"
```
Can you disable it then?

bryan-harter replied:
sms is not enabled by default

greglucas commented on sds_data_manager/lambda_code/SDSCode/cognito_signup_message.py:
```diff
@@ -0,0 +1,80 @@
+                + os.environ["COGNITO_DOMAIN"]
+                + "/login?client_id="
+                + command_line_client
+                + "&redirect_uri=https://example.com&response_type=code"
```
Why are we going to `example.com`?

bryan-harter replied:
We don't currently have anywhere else to redirect to, and unfortunately the sign-up URL requires a redirect_uri

greglucas commented on sds_data_manager/lambda_code/SDSCode/cognito_signup_message.py:
```diff
@@ -0,0 +1,80 @@
+"""
```
After reading through this, I'm wondering why we need a custom message at all? I think it is generally for formatting, but I don't see much formatting below, so should we just use the default and live with that? I see adding formatting as a follow-up ticket.

bryan-harter replied:
The custom message contains the sign-up link, but the default message doesn't

greglucas commented on sds_data_manager/lambda_code/SDSCode/upload_api.py:
```diff
@@ -12,13 +12,10 @@
 def _load_allowed_filenames():
     """
     Load the config.json file as a python dictionary.
-
```
There should be a blank line here before you get to the "return" section I think, same for all of these below. They shouldn't be squashed together.

bryan-harter replied:
Yeah sorry I need to re-format this section, I'm not sure why it got all weird

greglucas commented on sds_data_manager/sds_cognito_stack.py:
```diff
@@ -0,0 +1,131 @@
+    ) -> None:
+        super().__init__(scope, construct_id, **kwargs)
+
+        ########### INIT
```
```suggestion
```

greglucas commented on sds_data_manager/sds_cognito_stack.py:
```diff
@@ -0,0 +1,131 @@
+        ########### INIT
+        initial_user_context = initial_email
+
+        ########### IAM POLICIES
```
```suggestion
```
Your variable names are good already, I don't think we need comments reiterating the variable names.

greglucas commented on sds_data_manager/sds_cognito_stack.py:
```diff
@@ -0,0 +1,131 @@
+            resources=["*"],
+        )
+
+        ########### COGNITO
```
```suggestion
```

greglucas commented on sds_data_manager/sds_cognito_stack.py:
```diff
@@ -0,0 +1,131 @@
+            id_token_validity=cdk.Duration.minutes(60),
+            access_token_validity=cdk.Duration.minutes(60),
+            refresh_token_validity=cdk.Duration.minutes(60),
+            auth_flows=cognito.AuthFlow(
```
I'm not following what all of this does, but anything with admin always screams out to me. Should we only be using the srp workflow?

https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-authentication-flow.html#amazon-cognito-user-pools-admin-authentication-flow

greglucas commented on sds_data_manager/sds_cognito_stack.py:
```diff
@@ -0,0 +1,131 @@
+            prevent_user_existence_errors=True,
+        )
+
+        # Add aunique domain name where users can sign up / reset passwords
```
```suggestion
        # Add a unique domain name where users can sign up / reset passwords
```

greglucas commented on sds_data_manager/sds_cognito_stack.py:
```diff
@@ -0,0 +1,131 @@
+
+        # Add aunique domain name where users can sign up / reset passwords
+        # Users will be able to reset their passwords at:
+        # https://sds-login-{sds_id}.auth.us-west-2.amazoncognito.com/login?client_id={}&redirect_uri=https://example.com&response_type=code
```
```suggestion
        # https://sds-login-{sds_id}.auth.us-west-2.amazoncognito.com/login?client_id={}&redirect_uri=https://example.com&response_type=code
```
Should we get a unique domain name to use here? Each user/stack could get deployed to subdomains to keep them separate:
`stack1.imap-sdc.com`

bryan-harter replied:
I would say yes definitely, but Route53 stuff is probably another ticket

greglucas commented on sds_data_manager/sds_cognito_stack.py:
```diff
@@ -0,0 +1,131 @@
+                username=initial_user_context,
+            )
+
+        ########### LAMBDA
```
```suggestion
```

greglucas commented on sds_data_manager/sds_data_manager_stack.py:
```diff
@@ -15,7 +15,13 @@
+        construct_id: str,
+        sds_id: str,
+        userpool_id: str = "",
+        app_client_id: str = "",
```
```suggestion
        userpool_id: str
        app_client_id: str,
```
Can the ID be an empty string? Perhaps push the validation up a higher level and make this an argument without a default.

greglucas commented on tests/unit/test_sds_cognito_validity.py:
```diff
@@ -0,0 +1,16 @@
+    app = core.App(context={"SDSID": "unit-testing"})
+    sds_id = "".join([random.choice(string.ascii_lowercase) for i in range(8)])
+    stack = SdsCognitoStack(app, f"sds-cognito-{sds_id}", sds_id)
+    assertions.Template.from_stack(stack)
```
Can you add a test to actually assert what you want to verify for the Cognito service? I tried to put an example in #42 for how to do that.

Going forward we want to add robust tests for every infrastructure piece we are adding.

greglucas reviewed (commented):
I'm still not convinced we actually want to handle Cognito like this, I think we should be doing some architecture diagrams with this to demonstrate what we want before jumping into code. It seems very complicated for what should be a pretty stationary authentication feature we're adding. i.e. I view this as being basically standalone and deployed once/twice, perhaps a separate App entirely.

If others like this approach that is fine though and I'm happy to be let this go in :)

bryan-harter committed:
Update sds_data_manager/sds_data_manager_stack.py

Force the stack to have userpool_id and app_client_id, rather than provide empty defaults

Co-authored-by: Greg Lucas <greg.m.lucas@gmail.com>

bryan-harter committed:
Typo in last commit

bryan-harter committed:
Putting required variables into the unit test

bryan-harter commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

laspsandoval commented on sds_data_manager/sds_cognito_stack.py:
```diff
@@ -0,0 +1,131 @@
+        self.app_client_id = command_line_client.user_pool_client_id
+
+        ########### OUTPUTS
+        cdk.CfnOutput(self, "COGNITO_USERPOOL_ID", value=self.userpool_id)
```
I've been told before to try to not use CloudFormation directory since it is better to use higher level constructs that provide more abstraction, better semantics, and sensible defaults. https://bobbyhadz.com/blog/cfn-resources-aws-cdk. Just wondering what your thoughts are on using another method (if one is possible)?

bryan-harter replied:
As in, we shouldn't use the functions with the "Cfn" prefix on the CDK? I generally agree, though as far as I know there is no other way to output variables from a stack

laspsandoval replied:
Yea. I did a brief Google search and you are right. I think there isn't another way.

bryan-harter commented:
Its been way too long since this review, I'm gonna close it and redo it to fit with the current version of the code base
